packages/material-ui/src/MenuList/MenuList.js

```diff
@@ -6,66 +6,120 @@ import ReactDOM from 'react-dom';
 import warning from 'warning';
 import ownerDocument from '../utils/ownerDocument';
 import List from '../List';
+import getScrollbarSize from '../utils/getScrollbarSize';
+import { setRef } from '../utils/reactHelpers';
 
-class MenuList extends React.Component {
-  state = {
-    currentTabIndex: null,
-  };
+function resetTabIndex(list, selectedItem, setCurrentTabIndex) {
+  const currentFocus = ownerDocument(list).activeElement;
 
-  componentDidMount() {
-    this.resetTabIndex();
+  const items = [];
+  for (let i = 0; i < list.children.length; i += 1) {
+    items.push(list.children[i]);
   }
 
-  componentWillUnmount() {
-    clearTimeout(this.blurTimer);
+  const currentFocusIndex = items.indexOf(currentFocus);
+
+  if (currentFocusIndex !== -1) {
+    return setCurrentTabIndex(currentFocusIndex);
   }
 
-  setTabIndex(index) {
-    this.setState({ currentTabIndex: index });
+  if (selectedItem) {
+    return setCurrentTabIndex(items.indexOf(selectedItem));
   }
 
-  handleBlur = event => {
-    this.blurTimer = setTimeout(() => {
-      if (this.listRef) {
-        const list = this.listRef;
+  return setCurrentTabIndex(0);
+}
+
+const MenuList = React.forwardRef(function MenuList(props, ref) {
+  const { actions, children, className, onBlur, onKeyDown, disableListWrap, ...other } = props;
+  const [currentTabIndex, setCurrentTabIndex] = React.useState(null);
+  const blurTimeoutIDRef = React.useRef();
+  const listRef = React.useRef();
+  const selectedItemRef = React.useRef();
+
+  React.useImperativeHandle(actions, () => ({
+    focus: () => {
+      if (selectedItemRef.current) {
+        selectedItemRef.current.focus();
+        return;
+      }
+
+      if (listRef.current && listRef.current.firstChild) {
+        listRef.current.firstChild.focus();
+      }
+    },
+    getContentAnchorEl: () => {
+      if (selectedItemRef.current) {
+        return selectedItemRef.current;
+      }
+      return listRef.current.firstChild;
+    },
+    adjustStyleForScrollbar: (containerElement, theme) => {
+      // Let's ignore that piece of logic if users are already overriding the width
+      // of the menu.
+      if (
+        containerElement.clientHeight < listRef.current.clientHeight &&
+        !listRef.current.style.width
+      ) {
+        const scrollbarSize = `${getScrollbarSize(true)}px`;
+        listRef.current.style[
+          theme.direction === 'rtl' ? 'paddingLeft' : 'paddingRight'
+        ] = scrollbarSize;
+        listRef.current.style.width = `calc(100% + ${scrollbarSize})`;
+      }
+      return listRef.current;
+    },
+  }));
+
+  React.useLayoutEffect(() => {
```

Comment:
My next chunk of work will involve evaluating the use of `currentTabIndex` in detail and changing the approach. I would prefer to not change that notably in this pull request.

I think changing this `useLayoutEffect` to `useEffect` is doable. The main case to make sure I'm handling robustly is the auto-focus case for Menu -- I need to ensure that `currentTabIndex` is correct when the `focus` logic executes. This may require a ref for knowing whether or not we have already executed `resetTabIndex` for mount and if `focus` gets called first then go ahead and execute `resetTabIndex` (and then skip doing it in the effect).

Of course, assuming `Menu` eventually gets converted to a function component, the `useLayoutEffect` question will come up again since that would be the most natural way to convert its `componentDidMount` which executes the auto-focus logic.

Reply:
After looking at this more closely, I realized this was completely harmless to change from `useLayoutEffect` to `useEffect`. The tab index is **not** used for the auto-focus functionality, so it only matters once the user starts hitting the tab key, and for that purpose it is perfectly fine to wait for `useEffect`.